compiler/src/iree/compiler/Codegen/Common/Transforms.cpp

```diff
@@ -359,4 +359,342 @@ void populateSwapExtractWithExpandPattern(RewritePatternSet &patterns) {
+  // the loop, the offsets and sizes will be computed per reassociation group.
+  SmallVector<OpFoldResult> expandedOffsets, expandedSizes;
+  SmallVector<OpFoldResult> expandedStrides(srcShape.size(),
+                                            rewriter.getIndexAttr(1));
+
+  for (auto [collapsedSize, collapsedOffset, reassocIndices] :
+       llvm::zip_equal(collapsedSizes, collapsedOffsets,
+                       collapseShapeOp.getReassociationIndices())) {
+    // CASE #1 - size and/or offset are dynamic.
+    if (isa<Value>(collapsedSize) || isa<Value>(collapsedOffset)) {
+      // Special case especially for collapse shape of convolution filter in
+      // IGEMM, while the offset is dynamic and the size is static.
+      if (isa<Attribute>(collapsedSize) && isa<Value>(collapsedOffset) &&
+          reassocIndices.size() != 1) {
+        // Check if offset is from affine.apply of form (d0 * K) or (K * d0).
+        auto applyOp = collapsedOffset.dyn_cast<Value>()
+                           .getDefiningOp<affine::AffineApplyOp>();
+        if (!applyOp) {
+          return rewriter.notifyMatchFailure(sliceOp,
+                                             "offset is not from affine.apply");
+        }
+
+        AffineMap map = applyOp.getAffineMap();
+        if (map.getNumInputs() != 1 || map.getNumResults() != 1) {
+          return rewriter.notifyMatchFailure(
+              sliceOp, "affine.apply map is not of form (d0 * K)");
+        }
+
+        AffineExpr expr = map.getResult(0);
+        int64_t multiplier;
+        if (auto dimExpr = dyn_cast<AffineDimExpr>(expr)) {
+          multiplier = 1;
+        } else if (auto binOp = dyn_cast<AffineBinaryOpExpr>(expr)) {
+          if (!binOp || binOp.getKind() != AffineExprKind::Mul) {
+            return rewriter.notifyMatchFailure(
+                sliceOp, "affine.apply is not a multiplication");
+          }
+          auto constExpr = isa<AffineConstantExpr>(binOp.getLHS())
+                               ? dyn_cast<AffineConstantExpr>(binOp.getLHS())
+                               : dyn_cast<AffineConstantExpr>(binOp.getRHS());
+          if (!constExpr) {
+            return rewriter.notifyMatchFailure(
+                sliceOp, "affine.apply must multiply by a constant");
+          }
+          multiplier = constExpr.getValue();
+        } else {
+          return rewriter.notifyMatchFailure(sliceOp,
+                                             "unable to get the multiplier");
+        }
+
+        auto maybeStaticSize = getConstantIntValue(collapsedSize);
+        if (!maybeStaticSize) {
+          return rewriter.notifyMatchFailure(sliceOp,
+                                             "collapsed size must be static");
+        }
+        if (maybeStaticSize.value() % multiplier != 0) {
+          return rewriter.notifyMatchFailure(
+              sliceOp, "collapsed size is not divisible by offset multiplier");
+        }
```

Comment:
There is actually a nice utility of AffineExpr to check if an expression is a multiple of some constant. I think this can be simplified a lot with it:
```suggestion
        auto maybeStaticSize = getConstantIntValue(collapsedSize);
        if (!maybeStaticSize) {
          return rewriter.notifyMatchFailure(sliceOp,
                                             "collapsed size must be static");
        }
        if (!map.getResult(0).isMultipleOf(maybeStaticSize.value())) {
          return rewriter.notifyMatchFailure(
              sliceOp, "collapsed size is not divisible by offset multiplier");
        }
```

Reply:
wow, this is a useful utility! Thanks for your suggestion.